Add equality/inequality operators to Dataset and DataElement

scaramallion opened the pull request:
Fixes issue #284

darcymason commented:
Thanks for putting this code together.  Having proper equality testing does make a lot of sense.
The code generally looks good, but I have a few comments:
- is it actually necessary to compare member variables explicitly?  E.g. in a [stackoverflow answer](http://stackoverflow.com/a/25176504/1987276), they just compare `a.__dict__` to `b.__dict__` (for Dataset, I guess this means calling the base class ` dict.__eq__`).  I'm not sure if this would flow through for nested Datasets via Sequences, but in theory it would (probably an `__eq__` etc. is needed or Sequence as well).  It shouldn't matter, but it might be an opportunity to simplify the code, and if in future other attributes were added, it would automatically still be right.
- in addition to the above, I wonder if other attributes of Dataset should be compared, other than just the dicom tags and values.  FileDataset carries attributes like `is_little_endian`, `filename`, etc.  Those could be compared through the `__dict__`s of the class instances.  But maybe it is more reasonable to compare for Dataset only on the dicom values, I'm not sure.
- also from the stackoverflow answer, ideally a `__hash__` function should be added as well
- about the try/except block... is it necessary to catch and ignore any exceptions across all that code?  I'd be inclined to let them be raised, or at least have a more targeted set of exceptions that were caught
- the `return` in the last loop in the Dataset code does not look right - it will always return on the first iteration
- at line 300, looks like the comment should be 'got to this point with*out*...;

Just some thoughts.  Thanks again, though, this is great to have the code basically in place and especially including unit tests as well.

scaramallion commented:
No problem, I use pydicom a lot so its only fair to help out when I can.

I have to admit this is my first time doing an equality operator. Simply comparing `__dict__` makes much more sense and works easily for DataElement. However for Dataset the elements don't seem to be in the class `__dict__` unless the DataElement's `name/keyword` is not in the data dictionary:

```python
>>> from pydicom.dataset import Dataset
>>> d = Dataset()
>>> d.SOPInstanceUID = '1.2.3.4'
>>> d.SOPEustaceUID = '1.2.3.4.5'
>>> d.__dict__
{'SOPEustaceUID': '1.2.3.4.5', '_parent_encoding': 'iso8859'}
```
Happily `values()` works just as well. Looking at the updated code reminds me why I love python. I've also removed the inequality unit tests since they're just the negation of the equality ones and added some for FileDataset.

scaramallion commented:
Ah dang, forgot to check with python3

scaramallion commented:
OK, this should hopefully be better.

I've defined `__hash__ = None` as this seems to be the recommendation for [Python 2](https://docs.python.org/2/reference/datamodel.html#object.__hash__) and I don't get `TypeError` raised otherwise.

I've also added a unit test for FileDataset

scaramallion commented:
Was there something else you'd like me to do with this? I think I covered all your points.

darcymason commented:
Sorry for the delay, not a lot of time lately ... accepting the merge now.
